The many flavours of WMI post (#103)
PowerShell/Community-Blog: Posts/2022/08/many-wmi-flavours.md

$collection.Put_()
```
Updating and deleting.

```powershell
$collection = $services.Get("\\.\root\SMS\site_PS1:SMS_Collection.CollectionID='PS1000043'")
$collection.Properties_['Name'].Value = 'AwesomeDeviceCollection_NewName'
$collection.Put_()

# Deleting

$collection.Delete_()
```

## The CIM Cmdlets

When you just want to perform a WMI query to analyze data, and not necessarily interact with it, you cannot beat the CIM Cmdlets.
They are extremely fast, and provide unique tools like auto-complete for class and namespace names and easy class retrival with _Get-CimClass_.

### Querying

Performing queries with the CIM Cmdlets is very pleasant. One line does it all.

```powershell
$result = Get-CimInstance -Query "Select * From Win32_Process Where Name = 'powershell.exe'"
```

The parameters are very similar to the _Get-WmiObject_ ones, and can be used as follows.

```powershell
$result = Get-CimInstance -ClassName 'Win32_Process' -Filter "Name = 'powershell.exe'"
```

The _auto-complete_ feature in Visual Studio Code.

![Auto-Complete with CIM](Get-CimInstance_autoComplete.png)

### Calling a WMI Method

The CIM Cmdlets introduces a unique way of calling WMI Methods.

> The results of a CIM query are called _CimInstance(s)_, and you cannot call instance methods like you would with the other two options.
> Instead, you call another Cmdlet called _Invoke-CimMethod_.

Reviewer: ```suggestion
The results of a CIM query are called **CimInstances**, and you cannot call instance methods like you would with the other two options.
Instead, you call another Cmdlet called `Invoke-CimMethod`.
```

  
- Using the _WMI Scripting API_ is great when you have to build whole scripts to manage WMI. Once you have the _SWbemServices_ object you can work with pretty much anything else.  
It is also rewarding performance-wise, compared to the previous method, since you are working with the RCW interfaces directly. The _System.Management_ namespace will wrap these interfaces to provide abstraction.  
But this comes at a cost. If you want to retrieve single objects or perform queries to analyze data, this method can be a little annoying to work with.  
  
- The _CIM Cmdlets_ are number one on performance when querying multiple-object datasets. The _CimInstance_ object is great to work with, specially when combining with other known objects, like the _PSCustomObject_.  
In the other hand, calling methods is not as straight forward as on the previous methods. And to access methods like _Put()_, _Get()_ or _Delete()_ can be challenging.

## Bonus

Let's create our own Namespace under root, and implement a custom class!
We will use the _System.Management_ namespace, but now you can use what you learn to implement this using the other methods as well.

> ```powershell-interactive
> PS>_ $namespace = ([wmiclass]'root:__Namespace').CreateInstance()
> PS>_ $namespace.Name = 'ScriptingBlogCoolNamespace'
> PS>_ $namespace.Put()
> 
> Path          : \\.\root:__NAMESPACE.Name="ScriptingBlogCoolNamespace"
> RelativePath  : __NAMESPACE.Name="ScriptingBlogCoolNamespace"
> Server        : .
> NamespacePath : root
> ClassName     : __NAMESPACE
> IsClass       : False
> IsInstance    : True
> IsSingleton   : False
> 
> PS>_ $class = [wmiclass]::new('root\ScriptingBlogCoolNamespace', '', $null)
> PS>_ $class['__Class'] = 'CustomClass'
> PS>_ $class.Qualifiers.Add('Static', $true)
> PS>_ $class.Properties.Add('Source', 'Custom class with .NET!')
> PS>_ $class.Properties.Add('PropertyKey', 1)
> PS>_ $class.Properties['PropertyKey'].Qualifiers.Add('Key', $true) ## You need a Key property, otherwise WMI wouldn't be able to assemble the path of a new instance.
> PS>_ $class.Put()
> 
> Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass
> RelativePath  : CustomClass
> Server        : .
> NamespacePath : root\ScriptingBlogCoolNamespace
> ClassName     : CustomClass
> IsClass       : True
> IsInstance    : False
> IsSingleton   : False
> 
> PS>_ $instance = ([wmiclass]'root\ScriptingBlogCoolNamespace:CustomClass').CreateInstance()
> PS>_ $instance.Source = 'CustomInstance!'
> PS>_ $instance.Put()
> 
> Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
> RelativePath  : CustomClass.PropertyKey=1
> Server        : .
> NamespacePath : root\ScriptingBlogCoolNamespace
> ClassName     : CustomClass
> IsClass       : False
> IsInstance    : True
> IsSingleton   : False
> ```

Reviewer: ````suggestion
```powershell
$namespace = ([wmiclass]'root:__Namespace').CreateInstance()
$namespace.Name = 'ScriptingBlogCoolNamespace'
$namespace.Put()
```

```Output
Path          : \\.\root:__NAMESPACE.Name="ScriptingBlogCoolNamespace"
RelativePath  : __NAMESPACE.Name="ScriptingBlogCoolNamespace"
Server        : .
NamespacePath : root
ClassName     : __NAMESPACE
IsClass       : False
IsInstance    : True
IsSingleton   : False
```

```powershell
$class = [wmiclass]::new('root\ScriptingBlogCoolNamespace', '', $null)
$class['__Class'] = 'CustomClass'
$class.Qualifiers.Add('Static', $true)
$class.Properties.Add('Source', 'Custom class with .NET!')
$class.Properties.Add('PropertyKey', 1)
## You need a Key property, otherwise WMI wouldn't be able to assemble the path of a new instance.
$class.Properties['PropertyKey'].Qualifiers.Add('Key', $true)
$class.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass
RelativePath  : CustomClass
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : True
IsInstance    : False
IsSingleton   : False
```

```powershell
$instance = ([wmiclass]'root\ScriptingBlogCoolNamespace:CustomClass').CreateInstance()
$instance.Source = 'CustomInstance!'
$instance.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
RelativePath  : CustomClass.PropertyKey=1
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : False
IsInstance    : True
IsSingleton   : False
```
````

And just like that, we have our own Namespace, Class and Instance!  

> The results on _WmiExplorer_.

Reviewer: ```suggestion
The results on **WmiExplorer**.
```

They are extremely fast, and provide unique tools like auto-complete for class and namespace names and easy class retrival with `Get-CimClass`.

### Querying

Performing queries with the CIM Cmdlets is very pleasant. One line does it all.

```powershell
$result = Get-CimInstance -Query "Select * From Win32_Process Where Name = 'powershell.exe'"
```

The parameters are very similar to the `Get-WmiObject` ones, and can be used as follows.

```powershell
$result = Get-CimInstance -ClassName 'Win32_Process' -Filter "Name = 'powershell.exe'"
```

The _auto-complete_ feature in Visual Studio Code.

![Auto-Complete with CIM](Get-CimInstance_autoComplete.png)

### Calling a WMI Method

The CIM Cmdlets introduces a unique way of calling WMI Methods.
The results of a CIM query are called **CimInstances**, and you cannot call instance methods like you would with the other two options.
Instead, you call another Cmdlet called `Invoke-CimMethod`.

```powershell
$commandLine = 'powershell.exe -ExecutionPolicy Bypass -Command "Write-Output ''Howdy! From WMI!''; Read-Host"'
$result = Get-CimClass -ClassName 'Win32_Process'

> $output = $result | Invoke-CimMethod -MethodName 'Create' -Arguments @{ CommandLine = $commandLine }
> 
> # Or
> 
> $output = Invoke-CimMethod -ClassName 'Win32_Process' -MethodName 'Create' -Arguments @{ CommandLine = $commandLine }

Reviewer: ```suggestion
$params = @{
  MethodName = 'Create'
  Arguments = @{
    CommandLine = $commandLine
  }
}
$output = $result | Invoke-CimMethod @params

# Or
$params = @{
  ClassName = 'Win32_Process'
  MethodName = 'Create'
  Arguments = @{
    CommandLine = $commandLine
  }
}
$output = Invoke-CimMethod @params
```

```

And the result:

```powershell
Invoke-CimMethod -ClassName 'Win32_Process' -MethodName 'Create' -Arguments @{ CommandLine = $commandLine }
```

```Output
ProcessId ReturnValue PSComputerName
--------- ----------- --------------
    14932           0
```

Have a hard time remembering parameters? Me too! Luckily _auto-complete_ also works with them.

![Auto-Complete with method parameters](Invoke-CimMethod_autoComplete.png)

### Creating, Updating and Deleting a WMI Class Instance

If you used the old WMI Cmdlets before, this will look familiar.

```powershell

> $collection = New-CimInstance -Namespace 'root\SMS\site_PS1' -ClassName 'SMS_Collection' -Property @{ Name = 'AwesomeDeviceCollection'; LimitingCollectionID = 'PS1000042' }

Reviewer: ```suggestion
$params = @{
  Namespace = 'root\SMS\site_PS1'
  ClassName = 'SMS_Collection'
  Property = @{
    Name = 'AwesomeDeviceCollection'
    LimitingCollectionID = 'PS1000042'
  }
}
$collection = New-CimInstance @params
```

```

Updating and deleting.

```powershell
Set-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'" -Property @{ Name = 'AwesomeDeviceCollection_NewName' }

#Or

> $collection = Get-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
> $collection | Set-CimInstance -Property @{ Name = 'AwesomeDeviceCollection_NewName' }

Reviewer: ```suggestion
$params = @{
  Namespace = 'root\SMS\site_PS1'
  Query = "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
}
$collection = Get-CimInstance @params
$collection | Set-CimInstance -Property @{
  Name = 'AwesomeDeviceCollection_NewName'
}
```

#Deleting

> $collection = Get-CimInstance -Namespace 'root\SMS\site_PS1' -Query "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"

Reviewer: ```suggestion
$params = @{
  Namespace = 'root\SMS\site_PS1'
  Query = "Select * From SMS_Collection Where Name = 'AwesomeDeviceCollection'"
}
$collection = Get-CimInstance @params
```

RelativePath  : CustomClass
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : True
IsInstance    : False
IsSingleton   : False
```

```powershell
$instance = ([wmiclass]'root\ScriptingBlogCoolNamespace:CustomClass').CreateInstance()
$instance.Source = 'CustomInstance!'
$instance.Put()
```

```Output
Path          : \\.\root\ScriptingBlogCoolNamespace:CustomClass.PropertyKey=1
RelativePath  : CustomClass.PropertyKey=1
Server        : .
NamespacePath : root\ScriptingBlogCoolNamespace
ClassName     : CustomClass
IsClass       : False
IsInstance    : True
IsSingleton   : False
```

And just like that, we have our own Namespace, Class and Instance!  
The results on **WmiExplorer**.

![Namespace, Class and Instance](NamespaceManiputlation.png)

## Conclusion

If you made it to the end, hopefully now you have the right tool for the right job, regarding WMI.  
There is no winner, all of them are good in specific situations.  
What they all have in common is that they, together, will make you a better System Administrator.
  
Thank you for following along on this journey, and I see you next time!

## Useful links

> [System.Management](https://docs.microsoft.com/en-us/dotnet/api/system.management?view=dotnet-plat-ext-6.0)  
> [WMI Scripting API](https://docs.microsoft.com/en-us/windows/win32/wmisdk/scripting-api-for-wmi)  
> [CIM Cmdlets](https://docs.microsoft.com/en-us/powershell/module/cimcmdlets/?view=powershell-7.2)  
> [Runtime-Callable Wrapper](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/runtime-callable-wrapper)  
> [WMI Documentation](https://docs.microsoft.com/en-us/windows/win32/wmisdk/wmi-start-page)  

Reviewer: ```suggestion
- [System.Management](https://docs.microsoft.com/en-us/dotnet/api/system.management?view=dotnet-plat-ext-6.0)  
- [WMI Scripting API](https://docs.microsoft.com/en-us/windows/win32/wmisdk/scripting-api-for-wmi)  
- [CIM Cmdlets](https://docs.microsoft.com/en-us/powershell/module/cimcmdlets/?view=powershell-7.2)  
- [Runtime-Callable Wrapper](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/runtime-callable-wrapper)  
- [WMI Documentation](https://docs.microsoft.com/en-us/windows/win32/wmisdk/wmi-start-page)  
```
